boofuzz/primitives/float.py

```diff
@@ -0,0 +1,87 @@
+import random
+import struct
+
+from ..fuzzable import Fuzzable
+
+
+class Float(Fuzzable):
+    """Primitive that generates random float values within a specific range and with a fixed format.
+
+    :type name: str, optional
+    :param name: Name, for referencing later.
+    :type default_value: float
+    :param default_value: Value used when the element is not being fuzzed.
+    :type s_format: str, optional
+    :param s_format: Format of the float value on encoding, defaults to .1f
+    :type f_min: float, optional
+    :param f_min: Minimal float value that can be generated while fuzzing, defaults to 0.0
+    :type f_max: float, optional
+    :param f_max: Maximal float value that can be generated while fuzzing, defaults to 100.0
+    :type max_mutations: int, optional
+    :param max_mutations: Total number of mutations for this individual primitive, defaults to 1000
+    :type seed: int or str or bytes or bytearray
+    :param seed: Set random.seed() with the given seed for reproducible results
+    :type encode_as_ieee_754: bool, optional
+    :param encode_as_ieee_754: Encode the float value as IEEE 754 floating point
+    """
+
+    def __init__(
+        self,
+        name=None,
+        default_value=0.0,
+        s_format=".1f",
+        f_min=0.0,
+        f_max=100.0,
+        max_mutations=1000,
+        seed=None,
+        encode_as_ieee_754=False,
+        *args,
+        **kwargs
+    ):
+
+        super(Float, self).__init__(name=name, default_value=str(default_value), *args, **kwargs)
+
+        self.s_format = s_format
+        self.f_min = f_min
+        self.f_max = f_max
+        self.max_mutations = max_mutations
+        self.seed = seed
+        self.encode_as_ieee_754 = encode_as_ieee_754
+
+    def mutations(self, default_value):
+        last_val = None
+        if self.seed is not None:
+            random.seed(self.seed)
+
+        for i in range(self.max_mutations):
+            if i == 0:
+                current_val = default_value
+            else:
+                current_val = random.uniform(self.f_min, self.f_max)
+
+            str_format = "%" + self.s_format
+            current_val = str_format % float(current_val)
+
+            if last_val == current_val:
+                continue
+            last_val = current_val
+            yield current_val
+
+    def encode(self, value, mutation_context=None):
+        if self.encode_as_ieee_754:
+            value = float(value)
+            value = self.__float_to_integer(value)
+            value = self.__integer_to_binary(value)
+
+        return value.encode()
+
+    @staticmethod
+    def __float_to_integer(value):
+        return sum(b << 8 * i for i, b in enumerate(struct.pack("f", value)))
+
+    @staticmethod
+    def __integer_to_binary(value, bit_length=32):
+        return bin(value).replace("0b", "").rjust(bit_length, "0")
```

Comment:
Right now this return a string like '11000110001110111000000100000000'. Encoded in bytes that is many 0x30 and 0x31. However, I'd expect something like 0x3f800000 (32 bit big endian IEEE-754 representation of 1.0).
That could be archived by `struct.pack('>f', 1.0)`.

Maybe we should also add an `endian` argument to Float as we have it for other primitives.

Later on we could even add a `Double` primitive for 64bit IEEE doubles.

Reply:
To clarify what the function actually does: It gives you the **bitwise** IEEE-754 presentation of a float.

Example:
```python
import struct


def __float_to_integer(value):
    return sum(b << 8 * i for i, b in enumerate(struct.pack("f", value)))


def __integer_to_binary(value, bit_length=32):
    return bin(value).replace("0b", "").rjust(bit_length, "0")


def encode(value: float):
    value = __float_to_integer(value)
    value = __integer_to_binary(value)
    return value


print(encode(1.0))
# output: 00111111100000000000000000000000
```
This should work, you can check it online, for example with: https://www.h-schmidt.net/FloatConverter/IEEE754.html.
It gives me the same result. The corresponding hex value is, like you said, 0x3f800000.

But if you think that the bitwise representation is not useful, I will change that to hexadecimal, like you already suggested.
I will definitively add an endian argument because otherwise, the correct interpretation of the value will be very platform-independent.

Reply:
I have now changed the IEEE 754 float encoding to hex, as you have requested.
The Float primitive and s_float function have the endian arguments, which accepts 'big' for big-endian and 'little' for little-endian.

Reply:
One important thing to mention is, that the encoded IEEE floats do not have the '0x' prefix. Is that a problem?
If that's so, I will change this later.

Reply:
The example for the bitwise representation above does return the correct value, however the encoding/representation isn't the one we're looking for IMHO.
It returns a string "00111111100000000000000000000000", which after encoding to bytes returns b"00111111100000000000000000000000". But each character in this bytes object represents one byte, not one bit.
We can see that by converting bytes to hex-string:
```python
>>> b"00111111100000000000000000000000".hex()
'3030313131313131313030303030303030303030303030303030303030303030'
```
0x30 is the hex code for ASCII `0` and 0x31 is `1`.
So this bytes object isn't 32 bits long, but 32 bytes.

I haven't worked with IEEE floats in network protocols yet, but some quick research showed that Modbus TCP expects the 32 bit representation that `struct.pack('>f', 1.0)` returns.

> One important thing to mention is, that the encoded IEEE floats do not have the '0x' prefix. Is that a problem?

The 0x prefix is just a hint for hex representation, it's not part of the actual number. So no, we don't have to add that.

With the latest changes, it should now work as expected. Thanks for that.